### PR TITLE
simple malloc and free using mmap

### PR DESCRIPTION
**Pull request by MovingtoMars**

### CODE REVIEW

**Comment by andrewrk** [comment]
one gotcha with syscalls is that the man pages for e.g. mmap are for the c wrapper of the syscall, which uses errno. zig thinks that errno is a phenomenally stupid idea and the syscalls return the kernel syscall return value directly.

https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/tree/arch/x86/kernel/sys_x86_64.c#n86

The kernel doesn't actually return -1 on error; it returns a negative error code. So I think the code shouldn't check for equality with -1, it should test if the value is <= 0, then it's an error.

Note that casting 0 to a `?&u8` violates the type, and we'll probably have casting 0 to ?&u8 cause a panic in debug mode.

File: std/syscall.zig [hunk 1]
```diff
@@ -17,6 +31,28 @@ fn syscall3(number: isize, arg1: isize, arg2: isize, arg3: isize) -> isize {
...
+}
+
+fn syscall6(number: isize, arg1: isize, arg2: isize, arg3: isize, arg4: isize, arg5: isize, arg6: isize) -> isize {
+    asm volatile ("syscall"
+        : [ret] "={rax}" (-> isize)
+        : [number] "{rax}" (number), [arg1] "{rdi}" (arg1), [arg2] "{rsi}" (arg2), [arg3] "{rdx}" (arg3), [arg4] "{r10}" (arg4), [arg5] "{r8}" (arg5), [arg6] "{r9}" (arg6)
+        : "rcx", "r11")
+}
+
+pub fn mmap(address: isize, length: isize, prot: isize, flags: isize, fd: isize, offset: isize) -> isize {
```
Comment (andrewrk) [inline std/syscall.zig]: in the same way that read and write take a `&u8` I think that mmap should take a `&u8` and the casting should happen here in this function when it calls syscall6.

**Comment by MovingtoMars** [comment]
Actually, I don't think testing for <= 0 would work, as some valid addresses could have negative values. Give me a second...

File: std/mem.zig [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import "syscall.zig";
+import "std.zig";
+
+pub fn malloc(bytes: isize) -> ?&u8 {
+    if (bytes > 4096) {
```
Comment (andrewrk) [inline std/mem.zig]: I wonder what this number depends on. It would be interesting to read the source code of `getpagesize`.

Comment (andrewrk) [inline std/mem.zig]: oops, I meant `sysconf(_SC_PAGESIZE)` not `getpagesize`.

**Comment by MovingtoMars** [comment]
Should detect errors correctly now.

**Pull request merged**